### ISSUE TITLE
[Entitas:] Collector destructor is not called

### ISSUE THREAD

**Issue by sschmid** [sschmid/Entitas-CSharp#273]
Just check Entitas.Unity.ViualDebugging Collector Destructor Test Scene and it seems like the destructor is not called anymore. Investigate.

**Comment by sschmid** [comment]
Looks like it's related to c# events...
Need to investigate or implement own event system...

**Comment by JuDelCo** [comment]
C# events doesn't work ?

**Comment by KumoKairo** [comment]
C# events are considered as "hooks" to root objects and GC doesn't collect them on the collection phase if there's a link via an event subscription to some object's method. C# destructors (_finalizers_) are actually a bit not what one usually expects them to be, they are mostly used to release some unmanaged resources or do some strange black magic like "reviving" an object from the grasp of the GC. For example there's an option of hooking some static link to the current object in a destructor and GC won't collect it (although It was going to - that's why the descructor was called). A more convenient way of making some disposable objects in C# is implementing a IDisposable interface and doing things inside of the Dispose method instead of the finalizer

Is this bug present in current 0.36 version? Or is it in the develop brahch? I can also take a look, I've spent quite some time fixing memory leaks via C# events, static links and other root objects

**Comment by sschmid** [comment]
C# events do work, but after a really short investigation, I have the feeling that assigning methods to event with += retains the object somehow even if MyEvent = null (which should remove all event listeners).

If I don't assign listeners ([like here](https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Collector/Collector.cs#L53-L68)) it works. As soon as I assign methods, the object is not released, even when I do `group.OnEntityAdded = null` somewhere else.

Result -> memory leak

This only affects a specific edge cases, the normal use-case shouldn't be affected. Still would be nice to solve this

**Comment by T2RKUS** [comment]
aren't you supposed to use -= on the event.

**Comment by sschmid** [comment]
usually yes, but in this test the group is deactivated and should remove all connected listeners by setting event = null
https://github.com/sschmid/Entitas-CSharp/blob/develop/Plugins/Entitas.Unity.VisualDebugging/Assets/Tests/Manual%20Tests/Collector%20Destructor/CollectorDestructorController.cs#L16

**Comment by KumoKairo** [comment]
Does it happen on a device? Can you provide some steps to reproduce this?

**Comment by sschmid** [comment]
```csharp
    void Start() {
        Contexts.sharedInstance.SetAllContexts();
        var context = Contexts.sharedInstance.game;

        // Collector
        context.GetGroup(GameMatcher.Test).CreateCollector();
        
        // Entity
        _initialEntity = context.CreateEntity();
        _initialEntity.isTest = true; // Collector collects
        context.DestroyEntity(_initialEntity); // Destroy, removes isTest, removes from group, but still retained by collector since we never call c.ClearCollectedEntities()
        context.ClearGroups(); // Remove all eventhandlers
    }
	
    void Update() {
        var context = Contexts.sharedInstance.game;
        for (int i = 0; i < 5000; i++) { // create a lot of entities and wait for the collector destructor to be called. When it's called it releases the entity and the entity is ready for reuse.
            var e = context.CreateEntity();
            if(e == _initialEntity) {
                Debug.Log("Reusing entity!"); // Entity finally got released and is reused
                EditorApplication.isPlaying = false;
            }
        }
    }
```

This is the expected behaviour, but it seems like c# events still prevent collector from being destructed

**Comment by sschmid** [comment]
Either we find a fix or as a last resort a custom event system can also work and might even have less memory footprint

**Comment by bddckr** [comment]
Whenever you reference a  C# event (e.g. by doing `anInstance.myEvent`) you get back a reference to that *event*.
Assigning and unassigning via `+=` and `-=` works because the original field is not changed. The operations act on the returned event and those two operations are just some nice syntatic sugar for a subscribe and unsubscribe method.
Now if you do `anInstance.myEvent = null` from outside the class of `anInstance` you just set that returned event to null. You're not changing the *field* on the instance!

A solution is to write a public method *inside that class* that sets the event *field* to null and use that method from outside to null the field.

>It's worth being aware of what field-like events are actually doing - they're creating a variable and an event at the same time. Within the class, you end up referencing the variable. From outside, you reference the event.
Source: http://stackoverflow.com/a/153744/283482

**Comment by kdrzymala** [comment]
side note: I ended up rewriting c# events to a custom solution in one of our games. c# delegates are immutable so when you += or -= you're leaving garbage ( similar to string concatenation). The custom solution doesn't have to be anything fancy -> this works for me: https://gist.github.com/kdrzymala/c2a96df48bb3dcbc06f8f27c246b8b1d 
So rewriting it to something like we use could gain some performance.

**Comment by KumoKairo** [comment]
@kdrzymala or you can go with plain-old vanilla observer pattern with Interfaces, it's event less memory-and-GC-heavy because you still need to create those delegates to put them in the list of Actions. (It's created every time if you pass any "method name" as it is to Add or Remove methods

**Comment by KumoKairo** [comment]
I've investigated this issue and it seems that's there a small design flaw. This Collector.Deactivate thing relies almost solely on C# finalizers, which means that entities will be retained untill GC decides to garbage collect. But basically we can drop and clean once the entity is destroyed - it's a determined behaviour - release an Entity from the collector once it's destroyed. Moreover, all of the collected entities in the collector are retained until GC decides that it's no longer alive, which actually means that it's not a "one entity" issue - any collected entity won't be reusable until it's collector is garbage collected.
So that's what happens now-
we create a collector that collects all of the Test entities
We create an entity and add a Test component to it - an entity is retained by three objects - a "Game" context, a Group of test entities and a collector of test entities.
We destroy the entity, it's released by "Game" context and Group of test entities, but NOT released by Collector of test entitites. C# Events have really little to do here, it's the retain-release logic that needs fixing. I'll go check further

**Comment by KumoKairo** [comment]
So basically I think I pinned it down. Actual fix will require a more deep understanding of Entitas internals which I don't have yet, but here's the problem.

 Entitas-CSharp/Plugins/Entitas.Unity.VisualDebugging/Assets/Tests/Manual Tests/Collector Destructor/CollectorDestructorController.cs#L12
Creates a collector based on a group. This group handles only `GroupEvent.Added`, so https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Collector/Collector.cs#L52 will be called only once for the `group.OnEntityAdded` event. `group.OnEntityRemoved` is missing - we don't event get to subscribe on it. I don't know whether it's intentional that the group of `AllOf` fires only this one event or not.
Group has some "unused" method https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Group/Group.cs#L54 which calls https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Group/Group.cs#L126 directly which then calls that `OnEntityRemoved` event directly. This is actually the only path that calls that directly without passing somewhere else. That method is "unused" because my resharper could not find any usages. The method that gets used instead is this one https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Group/Group.cs#L88 
It calls `removeEntitySystemSilently` and returns `true` if an entity was removed, not calling that `OnEntityRemoved` event too.
But even if those events worked correctly, calling https://github.com/sschmid/Entitas-CSharp/blob/develop/Entitas/Entitas/Collector/Collector.cs#L91 the collector will still never release it's retained entity.

To recap - `OnEntityRemoved` event is not called by the group because it doesn't even know that it needs to call it. After fixing that, `Collector` needs to be taught to release it's entities without relying on it's finalizer, just releasing an entity once it's destroyed (removed)

So it's nothing to do with the C# events themselves - it's the whole logic that needs to be fixed a bit

**Comment by bddckr** [comment]
The problem with that test is:
The created collector manages a list of Entities it collected. `_initialEntity` is in that list.
`_initialEntity` is also retained by the collector. This adds the collector to `_initialEntity.owners`.

So both the collector and `_initialEntity` have a reference to each other. That's why the collector is never GC'ed.

A simple fix to destroy that circular dependency is to change the test's `Start` method to include
`collector.ClearCollectedEntities();` at the end.
This removes `_initialEntity` from the list and also releases it from the collector, which in turn removes the collector from `_initialEntity.owners`. :tada:

No idea if that behavior really is a bug in Entitas or just a bug in the test. I'm too tired to figure it out today. 😄

**Comment by bddckr** [comment]
No idea if that behavior of retaining the entities is needed at all.

Maybe the collector should implement `IDisposable` to signal the user "this thing needs proper cleanup, otherwise hell will break loose".

Can you think of a reason why the collector retains the collected entities until one calls the clear method? Why isn't the collector subscribing to the entity released event and removes it from it's list of collected entities?
I guess if you tell the collector to use `GroupEvent.Removed` (or `AddedOrRemoved`) you want to be notified about completely removed entities, too. Not just "this entity was removed from this collector because its components don't match the matcher anymore".
That would explain the need to retain entities until they're not "collected" anymore.

**Comment by sschmid** [comment]
`context.ClearGroups()` is removed. The test is invalid now. Will close
